### PR TITLE
Set up app for Expo SDK, add haptic feedback for long presses with Expo Haptics

### LINKED ISSUE #3576: Mobile - Add haptic feedback when long pressing messages
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Log in to e.cash
2. Navigate to a conversation
3. Long press a message

## Expected Result:
There should be some kind of feedback when long pressing a message. 

## Actual Result:
Action menu is displayed without any previous feedback. 

## Workaround:
No need, visual issue. 

## Platform:
Where is this issue occurring?

 Web
 iOS ✔️ 
 Android ✔️ 
 Desktop App
 Mobile Web 

**Version Number:** 1.0.68-3

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:**

https://user-images.githubusercontent.com/44479856/121866175-f939b400-ccfe-11eb-9bb4-4bc187964671.mp4

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

___ 
From @quinthar https://expensify.slack.com/archives/C01GTK53T8Q/p1623442462228100
> When long-tapping on mobile, there should be some haptic feedback (ie, a little "tap" feeling when the long-trap registers), like in Slack.

### PR DESCRIPTION
**Pull request by jasperhuangg**
_NOTE TO REVIEWERS:_
> Most of the diff comes from changes to `package-lock.json`, which was updated to support unimodules and expo. While this diff may seem extremely large, I personally think it's useful to allow ourselves to use Expo SDKs, since they provide an easy-to-use and well-supported SDK for a ton of important features that we may want to add later on. You can read more about other Expo SDKs [here](https://docs.expo.io/versions/latest/) by using the links in the sidebar.

### Details

I previously opened https://github.com/Expensify/Expensify.cash/pull/3586, which added haptic feedback from a haptics library that wasn't as popular as we would have liked, so @marcaaron suggested I use Expo Haptics. 

This PR sets up our app with `react-native-unimodules`, which is needed to use any Expo SDKs, including `expo-haptics`. I used [this guide](https://docs.expo.io/bare/installing-unimodules/) provided by Expo to set up `react-native-unimodules`, which is the source of the diffs in any Objective-C, Java, gradle, and Pod files.

It also adds haptic feedback using `expo-haptics` for any long-presses detected on `PressableWithSecondaryInteraction` components.

### Fixed Issues
Fixes https://github.com/Expensify/Expensify.cash/issues/3576

### Tests (must be tested on an actual device and NOT a simulator)
1. Open any chat with messages.
2. Long press a message.
3. Verify that you feel some haptic feedback (a very short and abrupt vibration).

### QA Steps
Same as the above tests

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
N/A

### CODE REVIEW

**Review by timszot — DISMISSED** [review]
LGTM, leaving for others to review

**Comment by jasperhuangg** [comment]
Just realized that this requires a change to the babel.config for web to work! I'm playing around with it now, will take it off draft-mode once I figured it out.

**Comment by jasperhuangg** [comment]
Got the webpack stuff working! Although, I've learned that we won't be able to use a lot of the Expo on web (and they'll throw exceptions if we try to).

**Comment by marcaaron** [comment]
>I've learned that we won't be able to use a lot of the Expo on web (and they'll actually throw exceptions if we try to)

@jasperhuangg is this ready for review? I wasn't sure based on the previous comment, but I wouldn't expect to be able to implement haptics on mobile web.

**Comment by jasperhuangg** [comment]
@marcaaron Yup! Sorry if that wasn't clear with my previous comment, this is all good to go 👍

**Review by marcaaron — CHANGES_REQUESTED** [review]
Works great on Android and doesn't break anything on web. I don't have an iOS device with haptics atm so I can't test that part, but everything builds OK.

I'm requesting changes because I believe we are possibly including packages that we don't need and can minimize the bundle size by removing them. If I'm super off about that lmk, but it would be good to confirm that's not the case.

**Comment by jasperhuangg** [comment]
Implemented the suggestions @marcaaron suggested, it still builds fine on Web, and haptics still works great on Android 👍

File: config/webpack/webpack.common.js [hunk 1]
```diff
@@ -20,6 +20,7 @@ const includeModules = [
     'react-native-modal',
     'react-native-onyx',
     'react-native-gesture-handler',
+    '@unimodules',
```
Comment (marcaaron) [inline config/webpack/webpack.common.js:23]: Do we still need this?

File: ios/Podfile.lock [hunk 2]
```diff
@@ -14,6 +14,21 @@ PODS:
   - boost-for-react-native (1.63.0)
   - CocoaAsyncSocket (7.6.5)
   - DoubleConversion (1.1.6)
+  - EXConstants (10.1.3):
```
Comment (marcaaron) [inline ios/Podfile.lock:17]: I think maybe we need to re-install iOS pods after the change I suggested. There are a lot of modules being added still.

**Review by marcaaron — CHANGES_REQUESTED** [review]

File: config/webpack/webpack.common.js (hunk 1 above)
Comment (jasperhuangg) [reply config/webpack/webpack.common.js:23]: Just tried it without it, and it builds fine on web, so I'll get rid of it.

File: ios/Podfile.lock [hunk 3]
```diff
@@ -717,6 +757,32 @@ EXTERNAL SOURCES:
...
+  UMImageLoaderInterface:
+    :path: "../node_modules/unimodules-image-loader-interface/ios"
+  UMPermissionsInterface:
+    :path: "../node_modules/unimodules-permissions-interface/ios"
+  UMReactNativeAdapter:
+    :path: "../node_modules/@unimodules/react-native-adapter/ios"
+  UMSensorsInterface:
+    :path: "../node_modules/unimodules-sensors-interface/ios"
+  UMTaskManagerInterface:
+    :path: "../node_modules/unimodules-task-manager-interface/ios"
```
Comment (jasperhuangg) [inline ios/Podfile.lock:785]: @marcaaron Running `pod install` didn't get rid of these dependencies, so I'm guessing bits and pieces of each of these may be necessary for `expo-haptics`, although I'm unsure as to why we'd need stuff like the BarCodeScannerInterface and the FaceDetectorInterface.

**Comment by jasperhuangg** [comment]
@marcaaron Got rid of the unnecessary stuff, should be good for another look!

File: ios/Podfile.lock (hunk 3 above)
Comment (marcaaron) [reply ios/Podfile.lock:785]: Cool thanks for double checking that.

**Review by marcaaron — APPROVED** [review]

**Comment by jasperhuangg** [comment]
@HorusGoul @timszot Would love to hear your thoughts on what we've done here! I'm definitely not the most knowledgable about this stuff.

**Comment by HorusGoul** [comment]
Have we considered supporting Mobile Web by leveraging the [Vibration API](https://developer.mozilla.org/en-US/docs/Web/API/Vibration_API)? Only Android browsers support that API, but we could implement it as a progressive enhancement, and when Apple decides to ship the API in Safari, it will work.

**Review by HorusGoul — COMMENTED** [review]
Have we considered supporting Mobile Web by leveraging the [Vibration API](https://developer.mozilla.org/en-US/docs/Web/API/Vibration_API)? Only Android browsers support that API, but we could implement it as a progressive enhancement, and when Apple decides to ship the API in Safari, it will work.

**Comment by jasperhuangg** [comment]
> Vibration API

Ohh interesting, I haven't heard about this yet. That would definitely be a great idea to get more coverage, interested on hearing other people's thoughts.

**Comment by marcaaron** [comment]
I'm cool with whatever you want to do here @jasperhuangg. 

I personally like the idea of including as many progressive web app features that we can into the mobile web side, but less certain if this is a company wide priority. Feels maybe a bit out of scope to me and we can probably go back and add something like this later.

**Comment by jasperhuangg** [comment]
> I'm cool with whatever you want to do here @jasperhuangg.
> 
> I personally like the idea of including as many progressive web app features that we can into the mobile web side, but less certain if this is a company wide priority. Feels maybe a bit out of scope to me and we can probably go back and add something like this later.

Sounds good, I think I'll drop a question in #product or #expensify-open-source, and depending on what they say I'll create a new issue for it. In the meantime, I'm gonna go ahead and merge this 👍

**Comment by mvtglobally** [comment]
@marcaaron Can you please confirm this is expected to be tested on Mobile devices only for haptic(Vibration) feedback? Desktop and Web will not be easy to validate

**Comment by marcaaron** [comment]
@mvtglobally That's correct. And also note that this will not work on mobile web either. This is expected.

Only Android and iOS devices with haptic feedback features + running the native app will pass this QA.

**Pull request merged**